src/rgw/driver/posix/rgw_sal_posix.cc

```diff
@@ -19,45 +19,156 @@
 #include <sys/xattr.h>
 #include <unistd.h>
 #include "rgw_multi.h"
-#include "rgw_acl_s3.h"
 #include "include/scope_guard.h"
 
 #define dout_subsys ceph_subsys_rgw
 #define dout_context g_ceph_context
 
 namespace rgw { namespace sal {
 
-const int64_t READ_SIZE = 8 * 1024;
+const int64_t READ_SIZE = 128 * 1024;
 const std::string ATTR_PREFIX = "user.X-RGW-";
 #define RGW_POSIX_ATTR_BUCKET_INFO "POSIX-Bucket-Info"
 #define RGW_POSIX_ATTR_MPUPLOAD "POSIX-Multipart-Upload"
 #define RGW_POSIX_ATTR_OWNER "POSIX-Owner"
+#define RGW_POSIX_ATTR_OBJECT_TYPE "POSIX-Object-Type"
 const std::string mp_ns = "multipart";
 const std::string MP_OBJ_PART_PFX = "part-";
 const std::string MP_OBJ_HEAD_NAME = MP_OBJ_PART_PFX + "00000";
 
+struct POSIXOwner {
+  rgw_user user;
+  std::string display_name;
+
+  POSIXOwner() {}
+
+  POSIXOwner(const rgw_user& _u, const std::string& _n) :
+    user(_u),
+    display_name(_n)
+    {}
+
+  void encode(bufferlist &bl) const {
+    ENCODE_START(1, 1, bl);
+    encode(user, bl);
+    encode(display_name, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::const_iterator &bl) {
+    DECODE_START(1, bl);
+    decode(user, bl);
+    decode(display_name, bl);
+    DECODE_FINISH(bl);
+  }
+  friend inline std::ostream &operator<<(std::ostream &out,
+                                         const POSIXOwner &o) {
+    out << o.user << ":" << o.display_name;
+    return out;
+  }
+};
+WRITE_CLASS_ENCODER(POSIXOwner);
+
+std::string get_key_fname(rgw_obj_key& key, bool use_version)
+{
+  std::string oid;
+  if (use_version) {
+    oid = key.get_oid();
+  } else {
+    oid = key.get_index_key_name();
+  }
+  std::string fname = url_encode(oid, true);
+
+  if (!key.get_ns().empty()) {
+    /* Namespaced objects are hidden */
+    fname.insert(0, 1, '.');
+  }
+
+  return fname;
+}
+
+static inline std::string gen_rand_instance_name()
+{
+  enum { OBJ_INSTANCE_LEN = 32 };
+  char buf[OBJ_INSTANCE_LEN + 1];
+
+#if 0
+  gen_rand_alphanumeric_no_underscore(driver->ctx(), buf, OBJ_INSTANCE_LEN);
+#else
+  static uint64_t last_id = UINT64_MAX;
+  snprintf(buf, OBJ_INSTANCE_LEN, "%lx", last_id);
```

Comment:
fmt::format()?

Reply:
Is fmt::format() better than snprintf()?  There's *lots* of C code in here that will have to be replaced when we switch to aio.